File: models/BirthDeath.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
This is an LNA++ application example with a simple birth-death process 
involving a constitutively active DNA, and stochastic production and 
degradation of mRNA and protein:

Model of birth-death process:
==============================
         k_m
DNA      -->  DNA+mRNA
         g_m
mRNA     -->  0
         k_p
mRNA     -->  mRNA + protein
         g_p
protein  -->  0

A more detailed model description is provided in 
documentation/Documentation.pdf.

This example is used for performance testing of LNA++. We 
1) show how to perform exact simulation of the stochastic process and use LNA++
to compute the mean and temporal auto-covariance of the proteins, 
2) compute sensitivities of the mean protein copy number with respect to the model 
parameters and compare the computed results to a finite difference approximation,
3) perform multi-start optimization in order to compare parameter inference 
with and without the analytical sensitivities and 
4) use LNA++ to evaluate the likelihood landscape for this model.
"""

# Open plot windows or plot to file?
plotToFile = False
import sys
if len(sys.argv) > 1 and sys.argv[1] == '--headless':
    plotToFile = True
    import matplotlib
    matplotlib.use('Agg')

import matplotlib.pyplot as plt

def plot(plotToFile, filename): 
    if plotToFile:
        plt.savefig(filename)
    else:
        plt.show()

# add LNA++ python module to search path
import os
scriptDir = os.path.dirname(os.path.realpath(__file__))
from sys import path
path.append(scriptDir + '/../python')


# Create the Python module for the birth / death system
from LNA import generateLNA
generateLNA(scriptDir + '/BirthDeath/BirthDeath.xml', 'BirthDeath', computeSS='BOTH')

# Import the generated module (adapt path first)
path.append(scriptDir + '/modules')
import BirthDeathLNA

import numpy as np

# model parameters
parameterNames =  ['$k_m$','$k_p$','$g_m$','$g_p$']
Theta = [20, 25, 10, 1]

# time span 0.0, 0.1, ... 10.0
tspan = np.arange(0, 10, 0.1).tolist()

# Simulate: IC = steady state
MRE,Var = BirthDeathLNA.LNA(Theta,tspan)

# plot results:
fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2)
fig.suptitle('Simulation: IC = Steady state')
# mean of mRNA
ax1.plot(tspan, MRE[0])
ax1.set_xlabel('Time')
ax1.set_ylabel('mRNA')
ax1.set_title('Macroscopic mean of mRNA')
# auto-covariance of mRNA
ax2.matshow(Var[0, 0, :, :], extent=[np.min(tspan), np.max(tspan), np.max(tspan), np.min(tspan)])
ax2.set_xlabel('Time')
ax2.set_ylabel('Time')
ax2.set_title('Autocovariance of mRNA')
# mean of protein
ax3.plot(MRE[1])
ax3.set_xlabel('Time')
ax3.set_ylabel('Protein')
ax3.set_title('Macroscopic mean of protein')
# auto-covariance of protein
ax4.matshow(Var[1, 1, :, :], extent=[np.min(tspan), np.max(tspan), np.max(tspan), np.min(tspan)])
ax4.set_xlabel('Time')
ax4.set_ylabel('Time')
ax4.set_title('Autocovariance of protein')
plot(plotToFile, 'BirthDeath1.pdf')



# Simulate: IC = steady state; observable = mRNA and protein
# Both species are observed (= mRNA,protein)
ObsIndex = [1, 2]
# Variance of measurement noise
VarNoise = [10.0,50.0]

# solve LNA and compute measured distribution
MRE, Var = BirthDeathLNA.LNA(Theta, tspan, merr =VarNoise,obsVar=ObsIndex)

# plot results:
fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2)
fig.suptitle('Simulation: IC = steady state; observable = mRNA and protein')
# mean of mRNA
ax1.plot(tspan, MRE[0])
ax1.set_xlabel('Time')
ax1.set_ylabel('mRNA')
ax1.set_title('Macroscopic mean of mRNA')
# auto-covariance of mRNA
ax2.matshow(Var[0, 0, :, :], extent=[np.min(tspan), np.max(tspan), np.max(tspan), np.min(tspan)])
ax2.set_xlabel('Time')
ax2.set_ylabel('Time')
ax2.set_title('Autocovariance of mRNA')
# mean of protein
ax3.plot(MRE[1])
ax3.set_xlabel('Time')
ax3.set_ylabel('Protein')
ax3.set_title('Macroscopic mean of protein')
# auto-covariance of protein
ax4.matshow(Var[1, 1, :, :], extent=[np.min(tspan), np.max(tspan), np.max(tspan), np.min(tspan)])
ax4.set_xlabel('Time')
ax4.set_ylabel('Time')
ax4.set_title('Autocovariance of protein')
plot(plotToFile, 'BirthDeath2.pdf')



# Simulate: IC = steady state; observable = protein
# observable is second species (= protein)
ObsIndex = [2]
# Variance of measurement noise
VarNoise = [50.0]

# solve LNA and compute measured distribution
MRE, Var = BirthDeathLNA.LNA(Theta, tspan, merr=VarNoise,obsVar=ObsIndex)

# plot results:
fig, (ax1, ax2) = plt.subplots(1, 2)
fig.suptitle('Simulation: IC = steady state; observable = protein')
# mean of protein
ax1.plot(tspan, MRE[0])
ax1.set_xlabel('Time')
ax1.set_ylabel('Protein')
ax1.set_title('Macroscopic mean of protein')
# auto-covariance of protein
ax2.matshow(Var[0, 0, :, :], extent=[np.min(tspan), np.max(tspan), np.max(tspan), np.min(tspan)])
ax2.set_xlabel('Time')
ax2.set_ylabel('Time')
ax2.set_title('Autocovariance of protein')
plot(plotToFile, 'BirthDeath3.pdf')



# Simulate: IC = no steady state; observable = protein
# observable is second species (= protein)
ObsIndex = [2]
# Variance of measurement noise
VarNoise = [0.0]
# initial values (E(mRNA),E(Protein))
MRE0 = [2, 200];
# initial co-variances (cov[mRNA,mRNA],cov(mRNA,Protein),cov(Protein,Protein))
Var0 = [0, 0, 0];

# solve LNA and compute measured distribution
MRE, Var = BirthDeathLNA.LNA(Theta, tspan, Y0=MRE0, V0=Var0, merr=VarNoise, obsVar=ObsIndex)

# plot results:
fig, (ax1, ax2) = plt.subplots(1, 2)
fig.suptitle('Simulation: IC = no steady state; observable = protein')
# mean of protein
ax1.plot(tspan, MRE[0])
ax1.set_xlabel('Time')
ax1.set_ylabel('Protein')
ax1.set_title('Macroscopic mean of protein')
# auto-covariance of protein
ax2.matshow(Var[0, 0, :, :], extent=[np.min(tspan), np.max(tspan), np.max(tspan), np.min(tspan)])
ax2.set_xlabel('Time')
ax2.set_ylabel('Time')
ax2.set_title('Autocovariance of protein')
plot(plotToFile, 'BirthDeath4.pdf')



# Simulate: IC = no steady state; observable = protein; sensitivity = 1st order

# solve LNA and compute measured distribution
MRE, Var, Sens_MRE, Sens_Var = BirthDeathLNA.LNA(Theta, tspan, Y0=MRE0, V0=Var0, merr=VarNoise, obsVar=ObsIndex, computeSens=True)

# plot results:
fig, axs = plt.subplots(2, len(Theta))
fig.suptitle('Simulate: IC = no steady state; observable = protein; sensitivity = 1st order')
for i in range(len(Theta)):
    (ax1, ax2) = axs[0][i], axs[1][i]
    # Sensitivity of mean of protein
    ax1.plot(tspan, Sens_MRE[0, i, :])
    ax1.set_xlabel('Time')
    ax1.set_ylabel('Sensitivity of mean')
    ax1.set_title(parameterNames[i])
    # Sensitivity of auto-covariance of protein
    ax2.matshow(Sens_Var[0, 0, :, :, i], extent=[np.min(tspan), np.max(tspan), np.max(tspan), np.min(tspan)])
    ax2.set_xlabel('Time')
    ax2.set_ylabel('Sensitivity of autocovariance')
    ax2.set_title(parameterNames[i])
plot(plotToFile, 'BirthDeath5.pdf')



# Simulate: IC = no steady state; observable = protein; sensitivity = 1st & 2nd order

# solve LNA and compute measured distribution
MRE, Var, Sens_MRE, Sens_Var,Sens2_MRE,Sens2_Var = BirthDeathLNA.LNA(Theta, tspan, Y0=MRE0, V0=Var0, merr=VarNoise, obsVar=ObsIndex, computeSens2=True)

# plot results:
fig, axs = plt.subplots(len(Theta), len(Theta))
fig.suptitle('Simulation of mean: IC = no steady state; observable = protein; sensitivity = 1st & 2nd order')
for i in range(len(Theta)):
    for j in range(len(Theta)):
        ax1 = axs[i][j]
        # Sensitivity of mean of protein
        ax1.plot(tspan, Sens2_MRE[0, i, j, :])
        ax1.set_xlabel('Time')
        ax1.set_ylabel('Sensitivity of mean')
        ax1.set_title('(%s, %s)' % (parameterNames[i], parameterNames[j]))
plot(plotToFile, 'BirthDeath6.pdf')

fig, axs = plt.subplots(len(Theta), len(Theta))
fig.suptitle('Simulation of autocovariance: IC = no steady state; observable = protein; sensitivity = 1st & 2nd order')
for i in range(len(Theta)):
    for j in range(len(Theta)):
        ax1 = axs[i][j]
        # Sensitivity of mean of protein
        ax1.matshow(Sens2_Var[0, 0, i, j, :, :], extent=[np.min(tspan), np.max(tspan), np.max(tspan), np.min(tspan)])
        ax1.set_xlabel('Time')
        ax1.set_ylabel('Time')
        ax1.set_title('(%s, %s)' % (parameterNames[i], parameterNames[j]))
plot(plotToFile, 'BirthDeath7.pdf')

File: models/BirthDeath.m
% Simple birth/death model
% (C) Justin Feigelman
% justin.feigelman@helmholtz-muenchen.de
% Helmholtz Zentrum Muenchen

%%%%%%%%%%%%%%% create the Matlab executable for the birth / death system %%%%%%%%%%%%%%% 
%% define the stoichiometric matrix
clear all

model = 'BirthDeath';

% define the stoichiometric matrix
% the model correspond to the system
%     k_m
% DNA -->  DNA+mRNA
%      g_m
% mRNA --> 0
%       k_p
% mRNA  -->  mRNA + protein
%         g_p
% protein -->  0

addpath('../matlab')
generateLNA('BirthDeath/BirthDeath.xml', 'BirthDeath', 'BOTH')

%%%%%%%%%%%%%%% run some simulations and show the results %%%%%%%%%%%%%%%     
%% set up the simulations

% define model parameters, in same order as above
Theta = [20, 25, 10, 1];

% define observation times for the simulation
tspan = 0:0.1:10;

% add the path to the mex file
addpath('BirthDeath/')

%% compute solution assuming steady state and no measurement error

% compute the solution
tic
[MRE, Var] = BirthDeath_LNA(Theta, tspan);
toc

% plot MRE
close all
figure('Name','BirthDeath LNA Simulation: Steady State')
subplot(121)
plot(tspan,MRE(2,:),'LineWidth',2) % just plot protein
xlabel('Time')
ylabel('Protein')
title('Protein trajectory')

% plot variance
subplot(122)
imagesc(squeeze(Var(2,2,:,:)))
xlabel('Time')
ylabel('Time')
title('Protein Autocovariance matrix')

%% specify an observed variable (Protein)
tic
[MRE, Var] = BirthDeath_LNA(Theta, tspan, 2); % outputs protein (species 2) only
toc

%% specify measurement error (same for each species)
tic
[MRE, Var] = BirthDeath_LNA(Theta, tspan, 2, 50); % mRNA and protein have measurement variance 50
toc

figure('Name', 'BirthDeath LNA Simulation: Added Measurement Error')
subplot(121)
plot(tspan,MRE,'LineWidth',2)
xlabel('Time')
ylabel('Protein')
title('Protein trajectory')
subplot(122)
imagesc(tspan,tspan, Var)
xlabel('Time')
ylabel('Time')
title('Protein Autocovariance matrix')

%% specify measurement error (each separately)
close all
tic
[MRE, Var] = BirthDeath_LNA(Theta, tspan, [1,2], [50, 100]); % measurement error 50 and 100 for mRNA and protein, respectively
toc

figure('Name','BirthDeath LNA Simulation: Added Measurement Error for All Species')
subplot(121)
plot(tspan,MRE(2,:),'LineWidth',2)
xlabel('Time')
ylabel('Protein')
title('Protein trajectory')
subplot(122)
imagesc(tspan,tspan, squeeze(Var(2,2,:,:)))
xlabel('Time')
ylabel('Time')
title('Protein Autocovariance matrix')

%% specify initial values, variances

y0 = [2, 200]; % inital values [mRNA, Protein]
v0 = [0, 0, 0 ]; % variances

tic
[MRE, Var] = BirthDeath_LNA(Theta, tspan, 2, 0, y0, v0);
toc

figure('Name', 'BirthDeath LNA Simulation: Specify Initial Conditions')
subplot(121)
plot(tspan,MRE,'LineWidth',2)
xlabel('Time')
ylabel('Protein')
title('Protein trajectory')
subplot(122)
imagesc(tspan,tspan, Var)
xlabel('Time')
ylabel('Time')
title('Protein Autocovariance matrix')

%% Compute 1st order sensitivities
tic
[MRE, Var, Sens_MRE, Sens_Var] = BirthDeath_LNA(Theta, tspan, 2, 0, y0, v0);
toc
%% Plot first order sensitivity plots: analytical
hSens1_MRE = figure('Name','BirthDeath First Order Sensitivities of the MRE');

set(gcf, 'PaperPosition', [0.25, 2.5, 6,6])
labels = {'k_m','k_p','g_m','g_p'};
for i=1:4
    subplot(2,2,i)
    plot(tspan, Sens_MRE(i,:), 'LineWidth', 3)
    ylabel(labels{i})
    xlabel('Time')
    set(gca,'FontSize',20)    
end

%% First order sensitivity of the variance
hSens1_Var = figure('Name','BirthDeath First Order Sensitivities of the Autocovariance');
set(gcf, 'PaperPosition', [0.25, 2.5, 8,6])
for i=1:4
    subplot(2,2,i)
    imagesc(tspan,tspan,Sens_Var(:,:,i))
    title(['Sensitivity ' labels{i}])
%     xlabel('Time (a.u.)')
%     ylabel('Time (a.u.)')
    set(gca,'FontSize',20)
    h = colorbar;
    h.Label.String = ['Sensitivity ' labels{i}];
end
% colorbar
%%
saveas(gcf, '~/Projects/LNA++/paper/FirstOrderVarSensitivityAnalytical.pdf')


%% compute finite difference approximation to first order sensitivities
% delta=0.0001;
delta=5.9948e-04
clear FD_MRE FD_Var

for i=1:4
    [MREa, Vara] = BirthDeath_LNA(Theta, tspan, 2,0, y0, v0);
    Thetab = Theta;
    Thetab(i) = Thetab(i)*(1+delta);
    [MREb, Varb] = BirthDeath_LNA(Thetab, tspan, 2,0, y0, v0);    
    
    FD_MRE(i,:)     = (MREb-MREa)/(Theta(i)*delta);
    FD_Var(i,:,:)   = (Varb-Vara)/(Theta(i)*delta);
end


%% comparison of the MRE first order sensitivities
figure(hSens1_MRE)
for i=1:4
    subplot(2,2,i)
    hold all
    plot(tspan, FD_MRE(i,:), 'o', 'MarkerSize', 2)
%     title(['Sensitivity ' labels{i}])
    xlabel('Time (a.u.)')
    set(gca,'FontSize',20)
end
legend('NorthEast', {'LNA++','F.D.'})
%%
saveas(gcf, '~/Projects/LNA++/paper/FirstOrderMRESensitivityComparison.pdf')
% print('~/Projects/LNA++/paper/FirstOrderMRESensitivityComparison.eps','-depsc')
saveas(gcf, '~/Projects/LNA++/paper/FirstOrderMRESensitivityComparison.fig')

%% plot finite difference first order var sensitivity

figure('Name', 'BirthDeath F.D. First Order Sensitivities of the Autocovariance')
set(gcf, 'PaperPosition', [0.25, 2.5, 8,6])
for i=1:4
    subplot(2,2,i)
    imagesc(tspan,tspan,squeeze(FD_Var(i,:,:)))
    title(['Sensitivity ' labels{i}])
%     xlabel('Time (a.u.)')
%     ylabel('Time (a.u.)')
    set(gca,'FontSize',20)
    h = colorbar;
    h.Label.String = ['Sensitivity ' labels{i}];
end

saveas(gcf, '~/Projects/LNA++/paper/FirstOrderVarSensitivityFD.pdf')

%% subtract the variance sensitivities (normalized)
figure('Name', 'BirthDeath First Order Sensitivities Comparison')
set(gcf, 'PaperPosition', [0.25, 2.5, 8,6])
for i=1:4
    subplot(2,2,i)
    imagesc(tspan,tspan, abs( (squeeze(FD_Var(i,:,:)) - squeeze(Sens_Var(:,:,i))) ./ squeeze(Sens_Var(:,:,i))), [0,1])
    title(['Sensitivity ' labels{i}])
%     xlabel('Time (a.u.)')
%     ylabel('Time (a.u.)')
    set(gca,'FontSize',20)
    colorbar
end

%% subtract the variance sensitivities (raw)
figure('Name', 'BirthDeath First Order Sensitivities Comparison')
set(gcf, 'PaperPosition', [0.25, 2.5, 8,6])
for i=1:4
    subplot(2,2,i)
    imagesc(tspan,tspan, squeeze(FD_Var(i,:,:)) - squeeze(Sens_Var(:,:,i)))
    title(['Sensitivity ' labels{i}])
%     xlabel('Time (a.u.)')
%     ylabel('Time (a.u.)')
    set(gca,'FontSize',20)
    colorbar
end

%% try different perturbation sizes, compute the error
delta=logspace(-5,-1,10);
clear FD_MRE FD_Var var_err

for k=1:10
    for i=1:4
        [MREa, Vara] = BirthDeath_LNA(Theta, tspan, 2,0, y0, v0);
        Thetab = Theta;
        Thetab(i) = Thetab(i)*(1+delta(k));
        [MREb, Varb] = BirthDeath_LNA(Thetab, tspan, 2,0, y0, v0);    

%         FD_MRE(i,:)     = (MREb-MREa)/(Theta(i)*delta);
        FD_Var(i,:,:)   = (Varb-Vara)/(Theta(i)*delta(k));
        var_err(i,k) = norm( (squeeze(FD_Var(i,:,:)) - squeeze(Sens_Var(:,:,i))));
        fprintf('.')
    end
end

%% plot errors
figure
plot(var_err', 'o-','LineWidth',3)
set(gca, 'YScale', 'log')
%% compute second order sensitivities

tic
[MRE, Var, Sens_MRE, Sens_Var, Sens2_MRE, Sens2_Var] = BirthDeath_LNA(Theta, tspan, 2, 0, y0, v0);
toc
%% compute second order sensitivities assuming steady state
tic
[MRE, Var, Sens_MRE, Sens_Var, Sens2_MRE_SS, Sens2_Var_SS] = BirthDeath_LNA(Theta, tspan, 2, 0);
toc
%% plot second order sensitivity of the MRE

hSens2_MRE = figure('Name', 'BirthDeath Second Order Sensitivities of the MRE');
labels = {'k_m','k_p','g_m','g_p','m_0','p_0'};
set(gcf,'PaperPosition',[0.25,2.5,6,6])
k=1;
for i=1:4
    for j=1:4
        subplot(4,4,k)
        plot(tspan, squeeze(Sens2_MRE(i,j,:)), 'LineWidth', 3)

%         xlabel('Time (a.u.)')
        if i==1
            title(labels{j})
        end        
        if j==1
            ylabel(labels{i}, 'FontWeight', 'bold');
        end
        set(gca,'FontSize',14)

        k=k+1;
    end
end
%% plot second order sensitivity of the Autocovariance
close all
figure('Name', 'BirthDeath Second Order Sensitivities of the Autocovariance')
k=1;
for i=1:4
    for j=1:4
        subplot(4,4,k)
                
        clims(i,j,1) = min(min(squeeze(Sens2_Var(i,j,:,:))));
        clims(i,j,2) = max(max(squeeze(Sens2_Var(i,j,:,:))));
        
        imagesc(tspan,tspan,squeeze(Sens2_Var(i,j,:,:))) % squeeze(clims(i,j,:))')
%         title(sprintf('Sensitivity(%s,%s)', labels{i}, labels{j}))
        if i==1
            title(labels{j})
        end        
%         if j==1
%             ylabel({labels{i},'Time (a.u.)'}, 'FontWeight', 'bold');
%         end
        set(gca,'FontSize',14)

%         ylabel('Time (a.u.)')
%         xlabel('Time (a.u.)','FontWeight','bold')
        k=k+1;
        colorbar
    end
end
%%
saveas(gcf,'~/Projects/LNA++/paper/Sens2Var.pdf')


%% compute second order finite difference 
clear FD_MRE FD_Var
% Theta = [10 10 2 1];
% y0 = [5 50];
Theta = [20 25 10 1];
y0 = [2 200];
% v0 = [10,1,500];
merr=0;
[MREa, Vara] = BirthDeath_LNA(Theta, tspan, 2, merr, y0, v0);
 
% deltaVec = logspace(-2,-1,10);
% produces reasonable results with this perturbation size
% deltaVec=10^-1.3;
deltaVec=10^-4
for d=1:length(deltaVec)
    delta=deltaVec(d);

    tic
    for i=1:4
        for j=1:4
            fprintf('.')
            if i==j
                Theta_f = Theta; Theta_b=Theta;
%                 % relative perturbation
%                 Theta_f(i) = Theta(i)*(1+delta);
%                 Theta_b(i) = Theta(i)*(1-delta);
                
                % absolute peturbation
                Theta_f(i) = Theta(i)+delta;
                Theta_b(i) = Theta(i)-delta;

                [MRE_f, Var_f] = BirthDeath_LNA(Theta_f, tspan, 2, merr, y0, v0);
                [MRE_b, Var_b] = BirthDeath_LNA(Theta_b, tspan, 2, merr, y0, v0);    

%                 FD_MRE(i,i,d,:)     = (MRE_f+MRE_b-2*MREa)/(Theta(i)*delta)^2;
%                 FD_Var(i,i,d,:,:)   = (Var_f+Var_b-2*Vara)/(Theta(i)*delta)^2;
                
                FD_MRE(i,i,d,:)     = (MRE_f+MRE_b-2*MREa)/(delta^2);
                FD_Var(i,i,d,:,:)   = (Var_f+Var_b-2*Vara)/(delta^2);
            else
                Theta_ff = Theta; Theta_f0 = Theta; Theta_0f = Theta; 
                Theta_b0 = Theta; Theta_0b = Theta; Theta_bb = Theta;

                Theta_fb = Theta; Theta_bf = Theta;
                
%                 % relative perturbation size
%                 Theta_ff([i j]) = Theta([i j])*(1+delta);           
%                 Theta_f0(i) = Theta(i)*(1+delta);
%                 Theta_0f(j) = Theta(j)*(1+delta);
% 
%                 Theta_b0(i) = Theta(i)*(1-delta);
%                 Theta_0b(j) = Theta(j)*(1-delta);
%                 Theta_bb([i j]) = Theta([i j])*(1-delta);
% 
%                 Theta_fb([i j]) = Theta([i j]).*[1+delta,1-delta];
%                 Theta_bf([i j]) = Theta([i j]).*[1-delta,1+delta];
                
                % absolute perturbation size
                Theta_ff([i j]) = Theta([i j])+delta;           
                Theta_f0(i) = Theta(i)+delta;
                Theta_0f(j) = Theta(j)+delta;

                Theta_b0(i) = Theta(i)-delta;
                Theta_0b(j) = Theta(j)-delta;
                Theta_bb([i j]) = Theta([i j])-delta;

                Theta_fb([i j]) = Theta([i j])+ delta*[1,-1];
                Theta_bf([i j]) = Theta([i j])+ delta*[-1,1];
                
                [MRE_ff, Var_ff] = BirthDeath_LNA(Theta_ff, tspan, 2, merr, y0, v0);
%                 [MRE_f0, Var_f0] = BirthDeath_LNA(Theta_f0, tspan, 2, merr, y0, v0);
%                 [MRE_0f, Var_0f] = BirthDeath_LNA(Theta_0f, tspan, 2, merr, y0, v0);                
% 
                [MRE_bb, Var_bb] = BirthDeath_LNA(Theta_bb, tspan, 2, merr, y0, v0);
%                 [MRE_b0, Var_b0] = BirthDeath_LNA(Theta_b0, tspan, 2, merr, y0, v0);
%                 [MRE_0b, Var_0b] = BirthDeath_LNA(Theta_0b, tspan, 2, merr, y0, v0);

                [MRE_fb, Var_fb] = BirthDeath_LNA(Theta_fb, tspan, 2, merr, y0, v0);
                [MRE_bf, Var_bf] = BirthDeath_LNA(Theta_bf, tspan, 2, merr, y0, v0);
                
%                 FD_MRE(i,j,d,:)     = (MRE_ff - MRE_f0 - MRE_0f + 2*MREa - MRE_b0 - MRE_0b + MRE_bb)/(2*Theta(i)*Theta(j)*delta^2);
%                 FD_Var(i,j,d,:,:)   = (Var_ff - Var_f0 - Var_0f + 2*Vara - Var_b0 - Var_0b + Var_bb)/(2*Theta(i)*Theta(j)*delta^2);
                
%               alternative:
%                % relative perturbation
%                 FD_MRE(i,j,d,:,:)   = (MRE_ff - MRE_fb - MRE_bf + MRE_bb)/(4*Theta(i)*Theta(j)*delta^2);
%                 FD_Var(i,j,d,:,:)   = (Var_ff - Var_fb - Var_bf + Var_bb)/(4*Theta(i)*Theta(j)*delta^2);

                % using absolute perturbation
                FD_MRE(i,j,d,:,:)   = (MRE_ff - MRE_fb - MRE_bf + MRE_bb)/(4*delta^2);
                FD_Var(i,j,d,:,:)   = (Var_ff - Var_fb - Var_bf + Var_bb)/(4*delta^2);        
                
                
                var_err2(d,i,j) = norm( squeeze(FD_Var(i,j,d,:,:)) - squeeze(Sens2_Var(i,j,:,:)) ); 
            end        
        end
        fprintf('\n')
    end
    toc

end

%% plot the second order F.D. MRE 
figure(hSens2_MRE)

k=1;
for i=1:4
    for j=1:4
        subplot(4,4,k)
        hold all
        plot(tspan, squeeze(FD_MRE(i,j,:)), 'o', 'MarkerSize', 2)       
        k=k+1;
    end
end

legend({'LNA++','F.D.'})
saveas(gcf,'~/Projects/LNA++/paper/Sens2MREComparison.pdf')

%% Plot Second order F.D. Variance

figure('Name','BirthDeath F.D. Second Order Sensitivities of the Autocovariance')
k=1;
d=3;
for i=1:4
    for j=1:4
        subplot(4,4,k)
        imagesc(tspan,tspan,squeeze(FD_Var(i,j,1,:,:)))%, squeeze(clims(i,j,:))')
          if i==1
            title(labels{j})
        end        
%         if j==1
%             ylabel({labels{i},'Time (a.u.)'}, 'FontWeight', 'bold');
%         end
        set(gca,'FontSize',14)
        colorbar
%         xlabel('Time (a.u.)','FontWeight','bold')
        k=k+1;
    end
end

%%
saveas(gcf,'~/Projects/LNA++/paper/Sens2Var_FD.pdf')

%% comparison of variance sensitivities
figure('Name','BirthDeath Second Order Sensitivities comparison (normalized)')
k=1;
for i=1:4
    for j=1:4
        subplot(4,4,k)
        imagesc(tspan,tspan, abs((squeeze(FD_Var(i,j,1,:,:)) - squeeze(Sens2_Var(i,j,:,:))) ./ squeeze(Sens2_Var(i,j,:,:))), [0,1] )
        if i==1
            title(labels{j})
        end        
%         if j==1
%             ylabel({labels{i},'Time (a.u.)'}, 'FontWeight', 'bold');
%         end
        set(gca,'FontSize',14)
        colorbar
%         xlabel('Time (a.u.)','FontWeight','bold')
        k=k+1;
    end
end

%% comparison of variance sensitivities unnormalized
figure('Name','BirthDeath Second Order Sensitivities comparison (raw)')
k=1;
for i=1:4
    for j=1:4
        subplot(4,4,k)
        imagesc(tspan,tspan, (squeeze(FD_Var(i,j,1,:,:)) - squeeze(Sens2_Var(i,j,:,:))))%  ./ squeeze(Sens2_Var(i,j,:,:)) )
        if i==1
            title(labels{j})
        end        
%         if j==1
%             ylabel({labels{i},'Time (a.u.)'}, 'FontWeight', 'bold');
%         end
        set(gca,'FontSize',14)
        colorbar
%         xlabel('Time (a.u.)','FontWeight','bold')
        k=k+1;
    end
end



%% show error of FD estimator for variance sens.

X = reshape(var_err2, 10, 16);

figure
plot(X, 'o-', 'LineWidth', 3)
set(gca, 'YScale', 'log')

%% compare FD error for different deltas
figure
for k=1:10
    subplot(3,4,k)    
    imagesc( squeeze(FD_Var(4,4,k,:,:)) - squeeze(Sens2_Var(4,4,:,:)))
end